cpp/include/rapidsmpf/buffer/pinned_memory_resource.hpp

```diff
@@ -0,0 +1,236 @@
+ *
+ * This class provides an interface for allocating and deallocating pinned
+ * (page-locked) host memory asynchronously using CUDA streams.
+ */
+class PinnedMemoryResource {
+    friend class PinnedHostBuffer;
+
+  public:
+    /**
+     * @brief Constructs a new pinned memory resource.
+     *
+     * @param pool The pinned memory pool to use for allocations.
+     */
+    PinnedMemoryResource(PinnedMemoryPool& pool);
+
+    /**
+     * @brief Destroys the pinned memory resource.
+     *
+     * Note: This does not deallocate memory that was allocated through this resource.
+     * All allocated memory should be explicitly deallocated before destruction.
+     */
+    ~PinnedMemoryResource();
+
+    /**
+     * @brief Allocates pinned memory asynchronously.
+     *
+     * @param bytes The number of bytes to allocate.
+     * @param stream_ref The CUDA stream to use for the allocation operation.
+     * @return A pointer to the allocated memory, or nullptr if allocation failed.
+     */
+    void* allocate_async(size_t bytes, const cuda::stream_ref stream_ref);
+
+    /**
+     * @brief Deallocates pinned memory asynchronously.
+     *
+     * @param ptr A pointer to the memory to deallocate.
+     * @param stream_ref The CUDA stream to use for the deallocation operation.
+     */
+    void deallocate_async(void* ptr, const cuda::stream_ref stream_ref);
+
+  private:
+    struct PinnedMemoryResourceImpl;
+    std::unique_ptr<PinnedMemoryResourceImpl> impl_;
+};
+
+/**
+ * @brief A buffer that manages pinned host memory.
+ *
+ * This class provides a convenient interface for managing pinned host memory
+ * buffers. It automatically handles allocation and deallocation through a
+ * PinnedMemoryResource and supports both empty buffer creation and buffer
+ * creation with data copying.
+ */
+class PinnedHostBuffer {
+  public:
+    /**
+     * @brief Constructs an empty pinned host buffer.
+     *
+     * @param size The size of the buffer in bytes.
+     * @param stream The CUDA stream to use for memory operations.
+     * @param p_resource Pointer to the memory resource to use for allocation and
+     * deallocation.
+     */
+    PinnedHostBuffer(
+        size_t size, cuda::stream_ref stream, PinnedMemoryResource* p_resource
+    );
+
+    /**
+     * @brief Constructs a pinned host buffer and copies data into it.
+     *
+     * @param data Pointer to the source data to copy.
+     * @param size The size of the data to copy in bytes.
+     * @param stream The CUDA stream to use for memory operations.
+     * @param p_resource Pointer to the memory resource to use for allocation and
+     * deallocation.
+     */
+    PinnedHostBuffer(
+        void const* data,
+        size_t size,
+        cuda::stream_ref stream,
+        PinnedMemoryResource* p_resource
+    );
+
+    /**
+     * @brief Destroys the pinned host buffer and waits for the associated stream to
+     * complete.
+     */
+    ~PinnedHostBuffer() noexcept;
+
```

Comment:
Since the dtor deallocate the buffer, we need to implement the move operators and set `data_ = nullptr`.

Reply:
Good point. Let me fix that.

Reply:
this is now fixed